### PR TITLE
Feat/artificial drives

### PR DESCRIPTION
**Pull request by thestonefox**

### CODE REVIEW

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalTransformDrive.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        /// </summary>
+        [Tooltip("The GameObject that contains the meshes for the control."), InternalSetting, SerializeField]
+        protected GameObject interactableMesh;
+        /// <summary>
+        /// The <see cref="TransformPositionDifferenceRotation"/> to drive the rotation of the control.
+        /// </summary>
+        [Tooltip("The TransformPositionDifferenceRotation to drive the rotation of the control."), InternalSetting, SerializeField]
+        protected TransformPositionDifferenceRotation rotationModifier;
+        /// <summary>
+        /// The <see cref="ArtificialVelocityApplier"/> apply artificial angular velocity to the control after releasing.
```
Comment (bddckr) [inline Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalTransformDrive.cs]: `to apply`

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalTransformDrive.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        /// </summary>
+        [Tooltip("The ArtificialVelocityApplier apply artificial angular velocity to the control after releasing."), InternalSetting, SerializeField]
```
Comment (bddckr) [inline Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalTransformDrive.cs]: `to apply`

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/DirectionalTransformDrive.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        /// The <see cref="TransformPropertyApplier"/> to automatically move the drive to a specific location.
+        /// </summary>
+        [Header("Reference Settings"), Tooltip("The TransformPropertyApplier to automatically move the drive to a specific location."), InternalSetting, SerializeField]
+        protected TransformPropertyApplier propertyApplier;
+        #endregion
+
+        /// <summary>
+        /// A float range of 0f, 0f.
+        /// </summary>
+        private readonly FloatRange nilLimit = FloatRange.Zero;
```
Comment (bddckr) [inline Prefabs/Interactions/Controllables/SharedResources/Scripts/DirectionalTransformDrive.cs]: `private static readonly`

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/DirectionalTransformDrive.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        /// <summary>
+        /// The position to move the drive to automatically.
```
Comment (bddckr) [inline Prefabs/Interactions/Controllables/SharedResources/Scripts/DirectionalTransformDrive.cs]: `The position to automatically move the drive to.`

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs [hunk 5]
```diff
@@ -52,12 +60,21 @@ public abstract class RotationalDrive : Drive<RotationalDriveFacade, RotationalD
...
 
         /// <summary>
         /// Calculates the location of the rotational hinge for the drive.
         /// </summary>
         /// <param name="newHingeLocation">The new local space for the hinge point.</param>
         public abstract void CalculateHingeLocation(Vector3 newHingeLocation);
+        /// <summary>
+        /// Attempts to apply the existing angular velocity back on to the rotation of the drive.
+        /// </summary>
+        /// <param name="multiplier">Multiplies the amount of angular velocity to be applied.</param>
```
Comment (bddckr) [inline Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs]: `The amount to multiply the angular velocity to be applied by.`

Or something... Hard to get this right 😄

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs [hunk 6]
```diff
@@ -67,5 +111,97 @@ protected override float CalculateValue(DriveAxis.Axis driveAxis, FloatRange lim
...
+            Vector3 currentEulerAngle = GetDriveTransform().localEulerAngles;
+            if (facade.DriveAxis == DriveAxis.Axis.XAxis && !currentEulerAngle.y.ApproxEquals(0f, 1f))
+            {
+                currentEulerAngle.x = currentEulerAngle.y - (currentEulerAngle.x > (circleDegrees * 0.5f) ? currentEulerAngle.x - circleDegrees : currentEulerAngle.x);
+            }
+            return currentEulerAngle;
+        }
+
+        /// <summary>
+        /// Calculates a multiplier based on the direction the rotation is travelling.
```
Comment (bddckr) [inline Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs]: `traveling`

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs [hunk 7]
```diff
@@ -67,5 +111,97 @@ protected override float CalculateValue(DriveAxis.Axis driveAxis, FloatRange lim
...
+            else if (pseudoRotation > DriveLimits.maximum)
+            {
+                GetDriveTransform().localRotation = Quaternion.Euler(-AxisDirection * DriveLimits.maximum);
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Calculates the velocity that the drive should automatically rotate the control.
```
Comment (bddckr) [inline Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs]: Add `with` to the end of the sentence?

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs [hunk 8]
```diff
@@ -67,5 +111,97 @@ protected override float CalculateValue(DriveAxis.Axis driveAxis, FloatRange lim
...
+        /// </summary>
+        /// <returns>The velocity to drive the control automatically.</returns>
```
Comment (bddckr) [inline Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs]: Add `with` to the end of the sentence?

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs (hunk 6 above)
Comment (thestonefox) [reply Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs]: https://dictionary.cambridge.org/dictionary/english/travelling

Comment (bddckr) [reply Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs]: >us usually traveling

The British English strikes again 😛

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalTransformDrive.cs (hunk 1 above)
Comment (thestonefox) [reply Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalTransformDrive.cs]: the original sentence doesn't make sense anyway, I'm going to change it to:

`The ArtificialVelocityApplier that applies artificial angular velocity to the control after releasing.`

Comment (bddckr) [reply Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalTransformDrive.cs]: 👍

File: Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs (hunk 6 above)
Comment (thestonefox) [reply Prefabs/Interactions/Controllables/SharedResources/Scripts/RotationalDrive.cs]: stupid spellings

**Pull request merged**